### ISSUE TITLE
error TS2451: Cannot redeclare block-scoped variable 'ngDevMode'

### ISSUE THREAD

**Issue by niccolascosta** [angular/angular#21670]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

node_modules/@angular/core/src/render3/ng_dev_mode.d.ts(9,11): error TS2451: Cannot redeclare block-scoped variable 'ngDevMode'.
node_modules/adv-view-js-comp-core/node_modules/@angular/core/src/render3/ng_dev_mode.d.ts(9,11): error TS2451: Cannot redeclare block-scoped variable 'ngDevMode'.
declare global  {
    const ngDevMode: boolean;
}
export declare const _ngDevMode: boolean;
export {};

## Environment

<pre><code>
Angular version: 5.2.1

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  Windows

Others:

</code></pre>

**Comment by Toxicable** [comment]
I'm sorry but we can't reproduce the problem following the instructions you provided.

If the problem still exists please open a new issue following [our submission guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-an-issue).

**Comment by ronnyek** [comment]
I was actually just running into this very same problem. I created a library with ng-packagr. I then used npm link to include this into a brand new angular/cli project (angular ^5.0.0).

As soon as I included the module from my library, I got this compilation error. The said library is massive, so including that one here wouldn't be feasible. I'll see if I can knock out a bare library and see if it still has same end result. Maybe its angular mismatch, as my library targets angular 4+

**Comment by crowmagnumb** [comment]
I also just ran into the same problem upgrading to 5.2.1. I have my setup a little funky because I'm working on getting ng-packagr working on a big library but have some issues. So in the meantime, I am (and have been for a long time now) using a "link" in my tsconfig.json file. For example,

```
{
    "compilerOptions": {
        "paths": {
            "MyLibrary/*": ["../myLibrary/src/*"]
        }
    }
}
```

And then in my app I can do an import like ...

```
import { MyComponent } from "MyLibrary/module";
```

... module.ts is the file that defines the module.

But since myLibrary is outside the tree of my main app I also have to keep an up-to-date package.json file in the library (which is fine because I will eventually have to anyway for the correctly packaged module).  All fine until this release and now I get ...

```
ERROR in [at-loader] ../myLibrary/node_modules/@angular/core/src/render3/ng_dev_mode.d.ts:9:11
    TS2451: Cannot redeclare block-scoped variable 'ngDevMode'.

ERROR in [at-loader] ./node_modules/@angular/core/src/render3/ng_dev_mode.d.ts:9:11
    TS2451: Cannot redeclare block-scoped variable 'ngDevMode'.
```

... as it apparently tries to read something twice that it expects only once. Hope this helps. I'm stuck at the moment but on to a week's vacation now. Hopefully resolved by my return.  :)

**Comment by PH-Lars** [comment]
I'm having the same problem with 5.2.1.

For me, downgrading tot 5.1.3 solved it for now.

`npm install @angular/common@5.1.3 @angular/compiler@5.1.3 @angular/compiler-cli@5.1.3 @angular/core@5.1.3 @angular/forms@5.1.3 @angular/http@5.1.3 @angular/platform-browser@5.1.3 @angular/platform-browser-dynamic@5.1.3 @angular/platform-server@5.1.3 @angular/router@5.1.3 @angular/animations@5.1.3 --save`

**Comment by steveblue** [comment]
I have a similar and maybe related problem with `ngDevMode`. Closure Compiler cannot bundle my browser module in ADVANCED_OPTIMIZATIONS because `ngDevMode` is not declared. I haven't tried making it an extern yet, will submit another issue if I can not get it working.

**Comment by jacobbuck72** [comment]
Downgrading solved the problem for me too.

**Comment by actra-gschuster** [comment]
Ran into the same issue today, in my case the problem is one of my libraries that had a direct dependency on `@angular/core`.
When installing my library in the actual ng-project it also created a node_modules folder there so I had two versions of `@angular/core` in the project - and both declare ngDevMode in the global scope resulting in the error message described in this issue.

**Comment by steveblue** [comment]
Making `ngDevMode` an extern worked for me with Closure Compiler

**Comment by denver-HJS** [comment]
I believe @actra-gschuster 's explanation and the suggestion to declare the *@angular* umbrella dependencies as `peerDependencies` provided in #20101 are related. After following both of those suggestions the error cleared up for me.

*Note:* I'm also using ng-packagr to create my npm package.

**Comment by legend1992** [comment]
@NgModule({
  imports: [
    **HttpModule**,
    InMemoryWebApiModule.forRoot(InMemoryDataService)
  ]
})
missing HttpModule will cause this issue

**Comment by crowmagnumb** [comment]
Thanks @PH-Lars, Falling back to 5.1.3 worked for me too. peerDependencies solution not valid for my situation.

**Comment by vicb** [comment]
Closing as no proper repro.
I would really help if someone could create a new issue with a proper repro. Thanks.

**Comment by crowmagnumb** [comment]
I'm trying to find the time to do so. If I manage it I will post it here.

**Comment by vicb** [comment]
Thanks @crowmagnumb ! (please no not post here but open a new issue as we do not monitor closed issues)

**Comment by crowmagnumb** [comment]
OK, created new [issue](https://github.com/angular/angular/issues/21925) with repo showing the error.

**Comment by mariohmol** [comment]
Hi @crowmagnumb could u solve it? this was working just now.. i updated something (cli? typescript?) and cant make it work back

https://github.com/mariohmol/ang-jsoneditor

**Comment by maxime1992** [comment]
As @actra-gschuster [pointed out](https://github.com/angular/angular/issues/21670#issuecomment-359752703) with multiple `@angular/core` resolutions, this was my issue.

And especially a dumb one because I'm using yarn resolution to make sure that this does not happen...
But I've made an upgrade for all angular libs and forgot to bump to version number in resolution.
Just updated it that too and it was good :+1:

**Comment by mariohmol** [comment]
Hi @maxime1992 thanks for answering.

Are u talking about peerDependencies? https://github.com/mariohmol/ang-jsoneditor/commit/5fa007f1c5aa1d5b0725b2b506fc6372c1abc8ad

I've tryed out many things.. cloned from the scratch.. bu cant make it work with latest angular-cli  

check it out  https://github.com/angular/angular/issues/21925#issuecomment-370112005

**Comment by maxime1992** [comment]
@mariohmol I'm not talking about peerDependencies.

Ex of what I'm talking about:

```json
{
  "name": "...",
  "scripts": {
    ...
  },
  "dependencies": {
    "@angular/core": "5.2.7",
    ...
  },
  "devDependencies": {
    "typescript": "2.7.2",
    ...
  },
  "resolutions": {
    "@angular/core": "5.2.7",
    "typescript": "2.7.2"
  }
}
```

Here, no matter what the sub dependencies rely on (for @angular/core + typescript), it'll always resolve to one and only version: The one I've written in "resolutions".

Note: This only work with Yarn as far as I'm aware.

**Comment by mariohmol** [comment]
Tested as you suggested and got the same error.

The problem happens when I use a new version for angular-cli. What is your global angular-cli version?

**Comment by mariohmol** [comment]
Thanks.. i upgraded to code to have this as well.
https://github.com/mariohmol/ang-jsoneditor/commit/14b879fd5afeffbae6a8d7ca057cffc3dea7c2d8

But what i I think was my problem, inside the `src/app/component`  i had a node_modules folder . 
Ex in this folder: https://github.com/mariohmol/ang-jsoneditor/tree/master/src/app/component

**Comment by glaucomorais** [comment]
I had this problem too, I just deleted the node_modules from my generated package and the app worked.

**Comment by fahedalh** [comment]
Thanks @PH-Lars, Falling back to 5.1.3 worked for me too.

**Comment by RJPalombo** [comment]
Falling back to 5.1.3 worked for me too.

**Comment by tyholby** [comment]
I fixed it by updating my typescript dependency to "typescript": "2.5.3".
(I'm on "@angular/core": "5.2.8")

**Comment by moniqx4** [comment]
I had a similar error in regard to rxjs, and the issue was because I had node_modules installed in my src directory, which was conflicting with the one that was being using.

So I had project/src/node_modules, and I also had project/node_modules . I had to remove the one, project/node_modules, and then take the one in project/src/node_modules and move it to /project/ .

After that, no more error, and compiled properly. I assume I somehow messed up my install, when I first was creating project. Currently, running angular 5.2.9, and typescript 2.7.2.

**Comment by tonysamperi** [comment]
Same here with 5.2.9...
I'll try updating to 5.2.11

**Comment by guido611** [comment]
I was running in the same issue using Angular: 6.0.9.

I was installing the [ngx-accordion-table](https://www.npmjs.com/package/ngx-accordion-table) version 1.0.10. But it seem to be not compatible with my Angular version.

Needed to uninstall it from the node_module of my angular project and the global one in my home-folder. Now it is working again.

**Comment by emersinlioglu** [comment]
I had the same problem, when i create a lib project in projects folder and execute "npm install" in the lib-project.

I solved this problem by deleting node_modules folder from the lib-project.

i think only one node_modules folder can exist in an angular-project.

**Comment by madhuni** [comment]
@actra-gschuster you mentioned that the library you were using was creating the another `node_modules` folder. I am facing the same issue with a library which is creating its own node_modules folder. Can you please suggest what is the resolution of the problem? How can we overcome this issue?

**Comment by actra-gschuster** [comment]
@madhuni  Specify the libraries in the "resolutions" field of your main package.json, then they won't be installed twice in different versions.

**Comment by TomasAndersen** [comment]
This is also a known problem to occur if you have several angular projects in the same repository.. if you miss on the imports in one of your components (eg when copying code from one project to the other) and actually tries to import something from @angular/core from the other project you get this error...

e.g
`import { Injectable } from '../../../../THE_OTHER_PROJECT_FOLDER/node_modules/@angular/core';`

instead of the correct

`import` { Injectable } from '@angular/core';`

**Comment by jgf5013** [comment]
I had this same error. Turns out one of my dependencies in node_modules had it's own node_modules folder. Not sure why it was there in the first place but deleting that nested folder resolved the issue.

**Comment by sreekanth-ayyavari** [comment]
> Ran into the same issue today, in my case the problem is one of my libraries that had a direct dependency on `@angular/core`.
> When installing my library in the actual ng-project it also created a node_modules folder there so I had two versions of `@angular/core` in the project - and both declare ngDevMode in the global scope resulting in the error message described in this issue.

But how u resolved this

**Comment by sdeering** [comment]
This fixed it for me. Run this for latest Angular as at 12/10/2018.

`npm install @angular/common@6.1.8 @angular/compiler@6.1.8 @angular/compiler-cli@6.1.8 @angular/core@6.1.8 @angular/forms@6.1.8 @angular/http@6.1.8 @angular/platform-browser@6.1.8 @angular/platform-browser-dynamic@6.1.8 @angular/platform-server@6.1.8 @angular/router@6.1.8 @angular/animations@6.1.8 --save`

**Comment by avatsaev** [comment]
Angular 7, same issue after upgrading from Angular 6 with `ng update`

**Comment by EduardsE** [comment]
> Angular 7, same issue after upgrading from Angular 6 with `ng update`

Did you manage to find a solution?

**Comment by avatsaev** [comment]
had to add this line in the main tsconfig 

![image](https://user-images.githubusercontent.com/518974/47460647-3d3eba00-d7df-11e8-9e15-c2d08dfee06e.png)

**Comment by sgruhier** [comment]
it doesn't fix it for me (ng 7)

**Comment by adornala** [comment]
@avatsaev solution worked for me. But I wonder why I need to do that for some machines, for some it's working fine. Strange!

**Comment by codeepic** [comment]
It happened for me, because we have two Angular apps in one repo powered by the same backend app.  Some of the components in both apps are named similarly and when using WebStorm intellisense it's easy to pull the component from another app into your app. And since another component references it's own local node_modules, you're importing angular/core twice.

**Comment by adornala** [comment]
Doesn't it mean. Doing a fresh npm || yarn install should fix it?

**Comment by codeepic** [comment]
@adornala No, because somewhere in app1 component you still reference app2 component. 
So when you import app2 cmp into app1, you also import with it app2 @angular/core package hence the error.

**Comment by avatsaev** [comment]
@codeepic I don't think that's the issue because it happened to me on this repo (which is a simple angular cli project): https://github.com/avatsaev/angular-contacts-app-example

you can see I had to add the line in the tsconfig file

**Comment by codeepic** [comment]
@adornala - that was the issue in my case, so I decided to drop in my 5 cents in case someone else has more than 1 Angular app in the same repo and mistakenly imports a component from another app.

**Comment by soroushchehresa** [comment]
I resolved this issue in Angular 7 by set manually `"paths": { "@angular/*": ["node_modules/@angular/*"] }` to tsconfig.json.

```
{
  ...
  "compilerOptions": {
    ...
    "paths": {
      "@angular/*": [
        "node_modules/@angular/*"
      ]
    }
  }
}
```

**Comment by przemek-lewicki** [comment]
So i ran into this problem yesterday. I had my Angular project and another NPM Package with angular as a dependency. Both Angulars had the same version (7.2.0). The error occurred while trying to build the main app.

My solution to this was to upgrade Angular in my project to the newest version (7.2.3). It looks like there was a conflict while both Angular versions were the same

**Comment by sumegha26** [comment]
It happened because the version of angular available in library is not matching with angular version in your project.

If you have your own custom library then just change its version to the version of angular in your project and re-publish it. It worked for me !

**Comment by amitcdubey** [comment]
> Ran into the same issue today, in my case the problem is one of my libraries that had a direct dependency on `@angular/core`.
> When installing my library in the actual ng-project it also created a node_modules folder there so I had two versions of `@angular/core` in the project - and both declare ngDevMode in the global scope resulting in the error message described in this issue.

but how to resolve this issue ?

**Comment by amitcdubey** [comment]
How to resolve this issue ,i understand the issue but how to resolve this because i need to update the version !!!!!

**Comment by paulo-jm** [comment]
My case was compilation problem I found the error when i gave ng build and saw the error. maybe you forgot add one dependence

**Comment by brenden-t-r** [comment]
I'm not exactly sure the cause in my case, but I have a project with two node_modules and clearing out both and re-running npm install did the trick.
